sonic-net/sonic-buildimage
# [VOQ][saidump] Install rdbtools into the docker base related containers.

#### Why did I do it?

To fix the issue: https://github.com/sonic-net/sonic-buildimage/issues/13561
The existing saidump use https://github.com/sonic-net/sonic-swss-common/blob/master/common/table_dump.lua script which loops the ASIC_DB more than 5 seconds and blocks other processes access.

This solution uses the Redis SAVE command to save the snapshot of DB each time and recover later, instead of looping through each entry in the table.

Related PRs:
https://github.com/sonic-net/sonic-utilities/pull/2972
https://github.com/sonic-net/sonic-buildimage/pull/16466
https://github.com/sonic-net/sonic-sairedis/pull/1288
https://github.com/sonic-net/sonic-sairedis/pull/1298

#### Work item tracking
Microsoft ADO (number only):

#### How did I do it?
To use the Redis-db SAVE option to save the snapshot of DB each time and recover later, instead of looping through each entry in the table and saving it.

    1. Updated dockers/docker-base-bullseye/Dockerfile.j2, install Python library rdbtools into the all the docker-base-bullseye containers.

    2. Updated sonic-buildimage/src/sonic-sairedis/saidump/saidump.cpp, add a new option -r, which updates the rdbtools's output-JSON files' format.

    3. To add a new script file: syncd/scripts/saidump.sh into the sairedis repo. This shell script does the following steps:

      For each ASIC, such as ASIC0,

      3.1. Config Redis consistency directory. 
      redis-cli -h $hostname -p $port CONFIG SET dir $redis_dir > /dev/null

      3.2. Save the Redis data.
      redis-cli -h $hostname -p $port SAVE > /dev/null

      3.3. Run rdb command to convert the dump files into JSON files
        rdb --command json $redis_dir/dump.rdb | tee $redis_dir/dump.json > /dev/null

      3.4.  Run saidump -r to update the JSON files' format as same as the saidump before. 
           Then we can get the saidump's result in standard output."
           saidump -r $redis_dir/dump.json -m 100

      3.5. Clear the temporary files.
       rm -f $redis_dir/dump.rdb
       rm -f $redis_dir/dump.json

    4. Update sonic-buildimage/src/sonic-utilities/scripts/generate_dump. To check the asic db size and if it is larger than ROUTE_TAB_LIMIT_DIRECT_ITERATION (with default value 24000) entries, then do with REDIS SAVE, otherwise, to do with old method: looping through each entry of Redis DB.


#### How to verify it
1) On T2 setup with more than 96K routes, execute CLI command -- generate_dump
2) No error should be shown
3) Download the generate_dump result and verify the saidump file after unpacking it.

#### Which release branch to backport (provide reason below if selected)




#### Tested branch (Please provide the tested image version)





#### Description for the changelog




#### Link to config_db schema for YANG module changes


#### A picture of a cute animal (not mandatory but encouraged)
